core/chainio/avs_writer.go

```diff
@@ -70,30 +73,77 @@ func NewAvsWriterFromConfig(baseConfig *config.BaseConfig, ecdsaConfig *config.E
 	txOpts := *w.Signer.GetTxOpts()
 	txOpts.NoSend = true // simulate the transaction
+
 	tx, err := w.RespondToTaskV2Retryable(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
 	if err != nil {
 		return nil, err
 	}
 
 	err = w.checkRespondToTaskFeeLimit(tx, txOpts, batchIdentifierHash, senderAddress)
 	if err != nil {
-		return nil, err
+		return nil, retry.PermanentError{Inner: err}
 	}
 
-	// Send the transaction
+	// Set the nonce, as we might have to replace the transaction with a higher gas price
+	txNonce := big.NewInt(int64(tx.Nonce()))
+	txOpts.Nonce = txNonce
 	txOpts.NoSend = false
-	txOpts.GasLimit = tx.Gas() * 110 / 100 // Add 10% to the gas limit
-	tx, err = w.RespondToTaskV2Retryable(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
-	if err != nil {
-		return nil, err
+	i := 0
+
+	respondToTaskV2Func := func() (*types.Receipt, error) {
+		// We bump only when the timeout for waiting for the receipt has passed
+		// not when an rpc failed
+		gasPrice, err := w.ClientFallback.SuggestGasPrice(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		bumpedGasPrice := utils.CalculateGasPriceBumpBasedOnRetry(gasPrice, gasBumpPercentage, gasBumpIncrementalPercentage, i)
+		// new bumped gas price must be higher than the last one (this should hardly ever happen though)
+		if gasPrice.Cmp(txOpts.GasPrice) > 0 {
+			txOpts.GasPrice = bumpedGasPrice
+		} else {
+			// bump the last price 10% to replace it.
+			bumpAmount := new(big.Int).Mul(txOpts.GasPrice, big.NewInt(10))
+			bumpAmount = new(big.Int).Div(bumpAmount, big.NewInt(100))
+			txOpts.GasPrice = new(big.Int).Mul(txOpts.GasPrice, bumpAmount)
+		}
+		if i > 0 {
+			onRetry(txOpts.GasPrice)
+		}
+
+		err = w.checkRespondToTaskFeeLimit(tx, txOpts, batchIdentifierHash, senderAddress)
+		if err != nil {
+			// We bump the fee so much that the transaction cost is more expensive than the batcher fee limit
+			return nil, retry.PermanentError{Inner: err}
+		}
+
+		tx, err = w.RespondToTaskV2Retryable(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
+		if strings.Contains(err.Error(), "reverted") {
+			return nil, err
+		}
+
+		receipt, err := utils.WaitForTransactionReceiptRetryable(w.Client, context.Background(), tx.Hash(), timeToWaitBeforeBump)
+		if receipt != nil {
+			return receipt, nil
+		}
+
+		// if we are here, it means we have reached the receipt waiting timeout
+		// we increment the i here to add an incremental percentage to increase the odds of being included in the next blocks
+		i++
+
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("transaction failed")
 	}
 
-	txHash := tx.Hash()
-
-	return &txHash, nil
+	return retry.RetryWithData(respondToTaskV2Func, 1000, 2, 0, 60, retry.MaxElapsedTime)
```

Comment:
`RetryWithData` function name is misleading, sounds like `Data` is actually input and not output

Reply:
I agree, but this sounds like something that should be done in #1304. Though, I don't mind refactoring it.